hypothesis-python/src/hypothesis/searchstrategy/strategies.py

```diff
@@ -624,6 +624,41 @@ def do_draw(self, data):
                 # As long as we consume data, we'll eventually pass or raise.
                 # But if we don't this could be an infinite loop.
                 assume(data.index > start_index)
+
+        # Special-case for SampledFrom strategy.  This is special even among
+        # strategies with a known and small(ish) list of possible elements,
+        # because it implements drawing directly as choice of an index.
+        # That means we can use the "make your own luck" trick explained in
+        # the shrinking guide to find even very rare allowed values.
+        from hypothesis.searchstrategy.lazy import LazyStrategy
+        from hypothesis.searchstrategy.misc import SampledFromStrategy
+
+        while isinstance(self.filtered_strategy, LazyStrategy):
+            self.filtered_strategy = self.filtered_strategy.wrapped_strategy
+
+        if isinstance(self.filtered_strategy, SampledFromStrategy):
+            # This is equivalent to calculating the allowed values, choosing one,
+            # and writing that index to the buffer - and we sometimes do just that.
+            # `speculative_index` allows us to skip processing part of the list.
+            # (how much depends on a complicated distribution - more when shrinking,
+            #  less when hard to find.  ~1/2 on average, but it's never average!)
+            draw_length = data.blocks.last_block_length * 8
+            speculative_index = cu.integer_range(
+                data, 0, max(len(self.filtered_strategy.elements) - 2, 0)
+            )
+            allowed = []
+            for idx, value in enumerate(self.filtered_strategy.elements):
+                if self.condition(value):
+                    allowed.append((idx, value))
+                    if len(allowed) > speculative_index:
+                        data.draw_bits(draw_length, forced=idx)
+                        return value
+            if allowed:
+                # If there are no allowed values, fall to the standard handler.
```

Comment:
I found it confusing that the first line in the `if allowed` branch is a comment describing what happens when the branch is *not* taken.